plutus-metatheory/test/TestNEAT.hs

```diff
@@ -1,105 +1,260 @@
+      Just _  -> return ()
+      Nothing -> throwError (CtrexTypeCheckFail tyG tmG)
 
-prop_checkKindSound :: TyProp
-prop_checkKindSound k _ tyQ = isSafe $ do
-  ty <- withExceptT GenErrorP tyQ
+-- try to typecheck a term
+prop_typecheck :: (Kind (), ClosedTypeG) -> ClosedTermG -> ExceptT TestFail Quote ()
+prop_typecheck (k , tyG) tmG = do
+  ty <- withExceptT GenError $ convertClosedType tynames k tyG
   tyDB <- withExceptT FVErrorP $ deBruijnTy ty
-  withExceptT AgdaErrorP $ case checkKindAgda (AlexPn 0 0 0 <$ tyDB) (deBruijnifyK (convK k)) of
+  tm <- withExceptT GenError $ convertClosedTerm tynames names tyG tmG
+  tmDB <- withExceptT FVErrorP $ deBruijnTerm tm
+  case checkTypeAgda (AlexPn 0 0 0 <$ tyDB) (AlexPn 0 0 0 <$ tmDB) of
     Just _  -> return ()
-    Nothing -> throwError ()
+    Nothing -> throwCtrex (CtrexTypeCheckFail tyG tmG)
 
-prop_normalizePreservesKind :: TyProp
-prop_normalizePreservesKind k _ tyQ = isSafe $ do
-  ty  <- withExceptT GenErrorP tyQ
-  tyDB <- withExceptT FVErrorP $ deBruijnTy ty
-  tyN <- withExceptT AgdaErrorP $ case normalizeTypeAgda (AlexPn 0 0 0 <$ tyDB) of
-    Just tyN -> return tyN
-    Nothing  -> throwError ()
-  withExceptT AgdaErrorP $ case checkKindAgda (AlexPn 0 0 0 <$ tyN) (deBruijnifyK (convK k)) of
+prop_runCK :: (Kind (), ClosedTypeG) -> ClosedTermG -> ExceptT TestFail Quote ()
+prop_runCK (k , tyG) tmG = do
+  tm <- withExceptT GenError $ convertClosedTerm tynames names tyG tmG
+  tmDB <- withExceptT FVErrorP $ deBruijnTerm tm
+  case runCKAgda (AlexPn 0 0 0 <$ tmDB) of
     Just _  -> return ()
-    Nothing -> throwError ()
+    Nothing -> throwCtrex (CtrexTermEvaluationFail tyG tmG)
 
--- the agda implementation throws names away, so I guess we need to compare deBruijn terms
-prop_normalizeTypeSound :: TyProp
-prop_normalizeTypeSound k tyG tyQ = isSafe $ do
-  ty <- withExceptT GenErrorP tyQ
-  tyDB <- withExceptT FVErrorP $ deBruijnTy ty
-  tyN1 <- withExceptT AgdaErrorP $ case normalizeTypeAgda (AlexPn 0 0 0 <$ tyDB) of
-    Just tyN -> return tyN
-    Nothing  -> throwError ()
-  ty1 <- withExceptT FVErrorP $ unDeBruijnTy tyN1
+prop_runTCK :: (Kind (), ClosedTypeG) -> ClosedTermG -> ExceptT TestFail Quote ()
+prop_runTCK (k , tyG) tmG = do
+  tm <- withExceptT GenError $ convertClosedTerm tynames names tyG tmG
+  tmDB <- withExceptT FVErrorP $ deBruijnTerm tm
+  case runTCKAgda (AlexPn 0 0 0 <$ tmDB) of
+    Just _  -> return ()
+    Nothing -> throwCtrex (CtrexTermEvaluationFail tyG tmG)
 
-  ty2 <- withExceptT GenErrorP $ toClosedType k (normalizeTypeG tyG)
-  return (ty1 == (AlexPn 0 0 0 <$ ty2))
+prop_run_plcCK_vs_CK :: (Kind (), ClosedTypeG) -> ClosedTermG -> ExceptT TestFail Quote ()
+prop_run_plcCK_vs_CK (k , tyG) tmG = do
+  tm <- withExceptT GenError $ convertClosedTerm tynames names tyG tmG
+  tmPlcCK <- withExceptT CkP $ liftEither $ evaluateCk mempty tm
+  tmDB <- withExceptT FVErrorP $ deBruijnTerm tm
+  tmCK <- withExceptT Ctrex $
```

Comment:
Quoting my comment on a previous PR:

> Here we run `withExceptT` sequentially and then lift everything into the concurrent `Cool` via `isSafe`. Would it make more sense to generalize `isSafe` to work over any error, call `isSafe` for each of the checks (without calling `withExceptT`) and then fold the results with `Data.Coolean.&&&` in order to parallelize the checks?

> On the other hand given that we expect the tests to succeed, there's not much point in optimizing the non-happy path as that would likely make the happy path less efficient.

> Maybe worth mentioning this in a Note somewhere.

Reply:
I have given this some thought but I think I must be missing something. The checks are rather like a pipeline as each step passes on something to the next (e.g., `convert term;convert type; check the type; check the term in the type; convert to DeBruijn; runCK; runCEK; compare results` etc). Some steps could be parallised but I don't see how to make this work with Data.Coolean.&&& as there is some essential threading of state going on.

It reminds me of the applicative do paper (https://simonmar.github.io/bib/papers/applicativedo.pdf) where the first example is:
```numCommonFriends :: Id → Id → Haxl Int
numCommonFriends x y = do
fx ← friendsOf x
fy ← friendsOf y
return (length (intersect fx fy))
```
and monadic threading makes the first two steps needlessly sequential.